core/modules/czar/Czar.cc

```diff
@@ -0,0 +1,328 @@
+      _uqFactory(), _clientToQuery(), _mutex() {
+
+    // set id counter to milliseconds since the epoch, mod 1 year.
+    struct timeval tv;
+    gettimeofday(&tv, nullptr);
+    const int year = 60*60*24*365;
+    _idCounter = uint64_t(tv.tv_sec % year)*1000 + tv.tv_usec/1000;
+
+    ccontrol::ConfigMap cm(_config);
+    std::string logConfig = cm.get("log.logConfig", "", "");
+    if (not logConfig.empty()) {
+        LOG_CONFIG(logConfig);
+    }
+
+    LOGF(_log, LOG_LVL_INFO, "creating czar instance with name %s" % czarName);
+    LOGF(_log, LOG_LVL_DEBUG, "czar config: %s" % util::printable(_config));
+
+    _uqFactory.reset(new ccontrol::UserQueryFactory(_config, _czarName));
+}
+
+std::vector<std::string>
+Czar::submitQuery(std::string const& query,
+                  std::map<std::string, std::string> const& hints) {
+
+    LOGF(_log, LOG_LVL_INFO, "new query: %s" % query);
+    LOGF(_log, LOG_LVL_INFO, "hints: %s" % util::printable(hints));
+
+    // get some info from hints
+    std::string clientId;
+    int threadId = -1;
+
+    auto hintIter = hints.find("client_dst_name");
+    if (hintIter != hints.end()) clientId = hintIter->second;
+
+    hintIter = hints.find("server_thread_id");
+    if (hintIter != hints.end()) {
+        try {
+            threadId = boost::lexical_cast<int>(hintIter->second);
+        } catch (boost::bad_lexical_cast const& exc) {
+            // Not fatal, just means we cannot associate query with particular
+            // client/thread and will not be able to kill it later
+        }
+    }
+
+    // this is atomic
+    uint64_t taskId = _idCounter++;
+    LOGF(_log, LOG_LVL_DEBUG, "taskId: %s" % taskId);
+
+    // make table names
+    auto taskIdStr = std::to_string(taskId);
+    std::string const resultName = _resultConfig.dbName + ".result_" + taskIdStr;
+    std::string const lockName = _resultConfig.dbName + ".message_" + taskIdStr;
+
+    std::vector<std::string> result(4);
+
+    // instantiate message table manager
+    MessageTable msgTable(lockName, _resultConfig);
+    try {
+        msgTable.lock();
+    } catch (std::exception const& exc) {
+        result[0] = exc.what();
+        return result;
+    }
+
+    // make new UserQuery
+    ccontrol::ConfigMap cm(hints);
+    std::string defDb = cm.get("db", "Failed to find default database, using empty string", "");
+    ccontrol::UserQuery::Ptr uq;
+    {
+        std::lock_guard<std::mutex> lock(_mutex);
+        uq = _uqFactory->newUserQuery(query, defDb, resultName, taskId);
+    }
+
+    // check for errors
+    auto error = uq->getError();
+    if (not error.empty()) {
+        result[0] = "Failed to instantiate query: " + error;
+        return result;
+    }
+
+    // start execution
+    LOGF(_log, LOG_LVL_DEBUG, "submitting new query");
+    uq->submit();
+
+    // spawn background thread to wait until query finishes to unlock
+    auto finalizer = [](ccontrol::UserQuery::Ptr uq, MessageTable msgTable) {
```

Comment:
Having both `UserQuery::join()` and finalizer probably does not make much sense. We could probably get rid of `join()` if we want UserQuery to handle everything internally, but then finalizer can become a part of UserQuery class itself. Mess. I say keep it as it is for now until we have more problems to solve in that area.
I'll add a comment, and I think it may be clearer if I pass those two objects in lambda capture explicitly by value.